ipld/amt/src/node.rs

```diff
@@ -552,6 +552,96 @@ where
 
         Ok((true, did_mutate))
     }
+
+    /// Iterates through the current node in the tree and all subtrees. `start_at` refers to the
+    /// global AMT index, before which no values should be traversed and `limit` is the maximum
+    /// number of leaf nodes that should be traversed in this subtree. `offset` refers the offset
+    /// in the global AMT address space that this subtree is rooted at.
+    #[allow(clippy::too_many_arguments)]
+    pub(super) fn for_each_while_ranged<S, F>(
+        &self,
+        bs: &S,
+        start_at: Option<u64>,
+        limit: Option<u64>,
+        height: u32,
+        bit_width: u32,
+        offset: u64,
+        f: &mut F,
+    ) -> Result<(bool, u64, Option<u64>), Error>
+    where
+        F: FnMut(u64, &V) -> anyhow::Result<bool>,
+        S: Blockstore,
+    {
+        let mut traversed_count = 0_u64;
+        match self {
+            Node::Leaf { vals } => {
+                let start_idx = start_at.map_or(0, |s| s.saturating_sub(offset));
+                let mut keep_going = true;
+                for (i, v) in (start_idx..).zip(vals[start_idx as usize..].iter()) {
+                    let idx = offset + i;
+                    if let Some(v) = v {
```

Comment:
NOTE: you could write this as:

```suggestion
                for (idx, v) in iter::zip(offset.., vals)
                    .skip(start_idx as usize)
                    // skip empty entries
                    .filter_map(|(i, v)| v.as_ref().map(|v| (i, v)))
                {
```